### PR TITLE
Initial support of Zksync2 payment

### PR DESCRIPTION
**Pull request by elek**
This branch (or relation chain) contains multiple modifications to support ZkSync2 payment. 

Half of the patches just make it easier to implement and test:

 * the implementation can be called directly from sub-commands (like `payer transfer`). 

Right the new zksync2 is tested with:

``` 
crybapy payer transfer ./safe1.key --node-address https://zksync2-testnet.zksync.dev --contract 0x787BE80167b052A317B5Af24ed6C0abFdafB96dB --type zksync2 --chain-id 280 --amount 1 --price 1 --destination 0x4C466C61cf8cEEa6c7365f6F6490fcF765deC6cD --max-gas 200000000000
```

This command creates an in-memory SQL database and imports the required transaction to the db before executing the transfer.

One missing part is the audit. Until now we didn't do proper audit of transaction, but with ZkSync2 we need more time between the submission of transaction and the confirmation (when tx is confirmed on ETH chain).

Therefore we need to improve the audit process and make it an important step in the workflow.

### CODE REVIEW

File: cmd/crybapy/cmd_payer_balance.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+			return checkCmd(doPayerBalance(config, args[0]))
+		},
+	}
+	RegisterFlags(cmd, &config.payerCommandConfig.PayerConfig)
+	return cmd
+}
+
+func doPayerBalance(config *payerBalanceConfig, spenderKeyPath string) error {
+
+	log, err := zap.NewDevelopment()
```
Comment (egonelbre) [inline cmd/crybapy/cmd_payer_balance.go:32]: shouldn't we use `zap.NewProduction` for the root logger?

File: pkg/zksync2/payer.go [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+	"storj.io/crypto-batch-payment/pkg/pipelinedb"
+	"storj.io/crypto-batch-payment/pkg/storjtoken"
+	"strings"
+)
+
+type Payer struct {
+	log              *zap.Logger
+	maxFee           *big.Int
+	wallet           *zksync2.Wallet
+	ethereumProvider interface{}
```
Comment (egonelbre) [inline pkg/zksync2/payer.go:30]: Sidenote, I've been enjoying using `any` instead of `interface{}`.

File: cmd/crybapy/cmd_zksync_balance_test.go [hunk 3]
```diff
@@ -8,12 +8,12 @@ import (
 )
 
 func Test_doZkSyncBalance(t *testing.T) {
-
+	t.Skip("Rrequires new deployed contract to the new test chain")
```
Comment (egonelbre) [inline cmd/crybapy/cmd_zksync_balance_test.go:11]: typo

File: pkg/zksync2/payer.go [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+	key *ecdsa.PrivateKey,
+	chainID int,
+	maxFee *big.Int) (*Payer, error) {
+
+	ethereumSigner, err := zksync2.NewEthSignerFromRawPrivateKey(key.D.Bytes(), int64(chainID))
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+
+	zkSyncProvider, err := zksync2.NewDefaultProvider("https://zksync2-testnet.zksync.dev")
```
Comment (egonelbre) [inline pkg/zksync2/payer.go]: Should this be using testnet?

File: pkg/pipelinedb/db.go [hunk 5]
```diff
@@ -43,11 +43,42 @@ func NewDB(ctx context.Context, path string) (*DB, error) {
 	return OpenDB(ctx, path, false)
 }
 
-func OpenDB(ctx context.Context, path string, readOnly bool) (_ *DB, err error) {
-	if _, err := os.Stat(path); err != nil {
-		return nil, err
+func NewInMemoryDB(ctx context.Context) (*DB, error) {
```
Comment (egonelbre) [inline pkg/pipelinedb/db.go:46]: nit: Maybe `OpenAndCreate` or something? Usually `New` implies that the func doesn't do anything besides internally configuring. However, this also does quite a few transactions. The func above probably needs updating as well.

**Review by egonelbre — COMMENTED** [review]
First pass looks good, but I'll need to review this again, because it's quite a bit of code.

**Review by egonelbre — COMMENTED** [review]
Found two issues, otherwise looks good.

File: cmd/crybapy/cmd_payer_balance.go (hunk 1 above)
Comment (elek) [reply cmd/crybapy/cmd_payer_balance.go]: 1959568937eacb0c2f3d1b2a1be068f3e088f131

File: go.mod [hunk 6]
```diff
@@ -3,51 +3,58 @@ module storj.io/crypto-batch-payment
 go 1.17
```
[the remark this answers is not in the document]
Comment (elek) [reply go.mod:3]: As we are the only user, and this is a CLI tool (not supposed to be used by other tools), we can bump it even to 1.20...

File: pkg/zksync2/payer.go [hunk 7]
```diff
@@ -0,0 +1,218 @@
+package zksync2
+
+import (
```
[the remark this answers is not in the document]
Comment (elek) [reply pkg/zksync2/payer.go:3]: I fixed this one file, but FTR: we don't enforce import order (and copyright) in this repository (see `make lint`).

(We can enable both, if we enable both the formatters, linters and bulk apply changes...)

**Review by egonelbre — APPROVED** [review]

**Comment by elek** [comment]
Merged manually with keeping the commit history...

**Pull request closed without merging**